### PR TITLE
Gisell: Firestore rules hardening and chat flow fixes

### PR DESCRIPTION
**Pull request by GTavarez**
What This PR Does
This PR hardens Firestore security rules and fixes related chat/profile flow issues found during validation, while preserving intended app behavior.

Firestore Security Updates:
- Prevent role escalation on user creation (block self-assignment of admin)
- Remove obsolete volunteer-based access path
- Keep connect-to-child flow, but restrict that update path to connection fields only
- Tighten draft message update permissions with explicit field allowlist + immutable field checks
- Restrict subscriptions access to owner/admin

Application Fixes
- Fix duplicate React key warning in conversation list by using stable thread key
- Fix duplicate React key warning in letters flow by upserting by message id instead of appending duplicates
- Remove duplicate setDoc call in account creation flow
- Improve profile image fallback behavior to reduce Storage 404 noise in letterhome flow

Validation Performed
Tested locally on http://localhost:3010:

- Login flow
- Letterhome load
- Discovery connect flow
- Draft autosave + refresh persistence
- Send flow (pending_review)
- Read/unread updates
- Profile update flow
Result: no Firestore permission-denied errors observed in tested workflows.

## Summary by CodeRabbit

* **Bug Fixes**
  * Removed a duplicate user write during account creation.
  * Prevented duplicate messages by merging sent messages when an ID already exists.
  * Eliminated redundant conversation fields to avoid inconsistent displays.

* **Improvements**
  * Safer profile-image selection with better fallbacks and graceful handling when avatars are missing.
  * Clarified avatar fetch error messaging.

* **Security & Infrastructure**
  * Added comprehensive Firestore security rules and configured the project to use them.

## Security Rules Gap List (for public-readiness decision)

This section captures the current Firestore security-rule gaps and the threshold we should meet before considering rules “public-ready”.

### Agreed context
- `connected_penpals_count` is a derived/indexing field intentionally used because Firestore client queries cannot filter by array size on `connected_penpals`.
- `subscriptions` and `notifications` are test collections and are excluded from this pass.

### Current gaps
1. **Child discovery data exposure**
   - Current `/users` discovery path can expose full child documents when listed.
   - Need public-safe shape separation (discoverable vs sensitive fields).

2. **Letterbox uniqueness enforcement**
   - Rules validate membership/connection but do not strictly guarantee one conversation identity per user pair.
   - Need deterministic identity/path enforcement.

3. **Letter field ownership / transition hardening**
   - Author/moderation paths need strict ownership boundaries and status-transition constraints.
   - Prevent sender-controlled edits to recipient/system-owned fields.

4. **`connected_penpals_count` update-path constraints**
   - Since this is a derived field, updates must be restricted to the intended connection flow only.
   - No arbitrary direct increments/decrements or manual rewrites outside approved connection mutation path.

5. **Schema/type validation completeness**
   - Enforce strict key/type validation on mutable collections to reduce malformed/poisoned writes.

6. **Role-model consistency**
   - Ensure role checks in rules match current product role model (no stale assumptions).

7. **Security test coverage**
   - Add emulator allow/deny tests for critical paths and edge cases.

### Public-ready threshold proposal
- close all critical/high gaps
- split discoverable child data into public-safe shape
- strict write schema/type validation on all mutable collections
- enforce single conversation identity per pair
- lock letter field ownership + status transitions
- align role checks with current model
- add emulator allow/deny tests for critical paths
## Security Rules Gap List (for public-readiness decision)

This section captures current Firestore security-rule gaps and the threshold we should meet before considering rules “public-ready”.

### Agreed context
- `connected_penpals_count` is a derived/indexing field intentionally used because Firestore client queries cannot filter by array size on `connected_penpals`.
- `subscriptions` and `notifications` are test collections and are excluded from this pass.

### Current gaps
1. **Child discovery data exposure**
   - `/users` discovery can expose full child documents when listed.
   - Need public-safe shape separation (discoverable vs sensitive fields).

2. **Letterbox uniqueness enforcement**
   - Rules validate membership/connection but do not strictly guarantee one conversation identity per user pair.
   - Need deterministic identity/path enforcement.

3. **Letter field ownership / transition hardening**
   - Author/moderation paths need strict ownership boundaries and status-transition constraints.
   - Prevent sender-controlled edits to recipient/system-owned fields.

4. **`connected_penpals_count` update-path constraints**
   - Since this is a derived field, updates must be restricted to intended connection flow only.
   - No arbitrary direct increments/decrements or manual rewrites outside approved mutation path.

5. **Schema/type validation completeness**
   - Enforce strict key/type validation on mutable collections to reduce malformed writes.

6. **Role-model consistency**
   - Ensure role checks in rules match current product role model.

7. **Security test coverage**
   - Add emulator allow/deny tests for critical paths and edge cases.

### Public-ready threshold proposal
- close all critical/high gaps
- split discoverable child data into public-safe shape
- strict write schema/type validation on all mutable collections
- enforce single conversation identity per pair
- lock letter field ownership + status transitions
- align role checks with current model
- add emulator allow/deny tests for critical paths

### CODE REVIEW

File: firebase.json
Comment (madarez) [inline firebase.json]: Does your adding the rules here remain in sync with the backend with every merge? If you need to make sure it gets reflected through a github action, you may need to update `.github/workflows/ci-deploy.yml` too.

File: app/letterhome/page.js
[the remark this answers is not in the document]
Comment (GTavarez) [reply app/letterhome/page.js]: Good point, agreed. I switched this back to storage-based profile image loading (getUserPfp) and removed direct reliance on photo_uri for the active flow. I pushed the update.

File: app/utils/letterboxFunctions.js [hunk 1]
```diff
@@ -254,15 +259,14 @@ export const fetchRecipients = async (id) => {
...
-      members.push({ ...selUser.data(), id: selectedUserDocRef.id, pfp: selUser.photo_uri });
-    }
+    const userData = selUser.data() || {};
+    const downloaded = await getUserPfp(user.id);
+    members.push({
+      ...userData,
+      id: selectedUserDocRef.id,
+      pfp: downloaded || "",
+      photo_uri: downloaded || userData.photo_uri || "",
+    });
```
[the remark this answers is not in the document]
Comment (madarez) [reply app/utils/letterboxFunctions.js]: @GTavarez

File: firestore.rules [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+            request.resource.data.created_at == resource.data.created_at &&
+            request.resource.data.diff(resource.data).affectedKeys().hasOnly([
+              'status',
+              'content',
+              'updated_at',
+              'drafted_at',
+              'attachments',
+              'unread',
+              'deleted'
+            ]) &&
```
[the remark this answers is not in the document]
Comment (GTavarez) [reply firestore.rules]: Good catch, fixed.
I removed unread from isAuthorStatusUpdate() so sender status transitions can’t also mutate read state. Unread changes now stay isolated to isSettingRead() only. Rules updated and deployed to penpalmagicapp.

File: app/utils/letterboxFunctions.js [hunk 3]
```diff
@@ -287,3 +298,75 @@ export const sendLetter = async (letterData, letterRef, draftId) => {
     return false;
   }
 };
+
+export const createConnection = async (userDocRef, kidDocRef) => {
```
Comment (madarez) [inline app/utils/letterboxFunctions.js]: changes here not necessary...Are you mostly trying to format codes here?

File: firestore.rules [hunk 4]
```diff
@@ -72,14 +93,53 @@ service cloud.firestore {
...
+      allow list: if isSignedIn() && isAdmin();
+      allow update: if isSignedIn() &&
+        (
+          isAdmin() ||
+          (request.auth.uid == userId && isSafeSelfUpdate() && isUpdatedUserDataValid()) ||
+          (isListingChildren() && isConnectionOnlyUpdate() && isValidChildConnectionUpdate() && isUpdatedUserDataValid())
+        );
+    }
+
+    match /users_public/{userId} {
```
Comment (madarez) [inline firestore.rules]: User information should not be public.

Comment (GTavarez) [reply firestore.rules]: Acknowledged. This comment was on an outdated diff; the users_public change has been reverted and is not part of the current PR scope.
